### PR TITLE
added logging to help with perf debugging.

### PR DESCRIPTION
**Pull request by jean-philippe-martin**
This code adds log statements with "START" and "END" keywords so that we can then process the logs and see (i) how many times a specific operation was called, (ii) how long the operation took (per call, or in aggregate).

In addition it has the notion of intermediate "steps" so we can drill down individual operations.

### CODE REVIEW

File: src/main/java/org/broadinstitute/hellbender/dev/pipelines/bqsr/BaseRecalibratorDataflowUtils.java [hunk 1]
```diff
@@ -173,6 +185,16 @@ private static String posKeyEnd(final Locatable loc) {
         .of(
             new DoFn<Read, KV<String, Read>>() {
                 private static final long serialVersionUID = 1L;
+                private BunnyLog bunny;
```
Comment (davidadamsphd) [inline src/main/java/org/broadinstitute/hellbender/dev/pipelines/bqsr/BaseRecalibratorDataflowUtils.java]: Why logger above and bunny here?

File: src/main/java/org/broadinstitute/hellbender/dev/pipelines/bqsr/BaseRecalibratorDataflowUtils.java [hunk 2]
```diff
@@ -313,12 +353,22 @@ public void finishBundle(DoFn<KV<String, CoGbkResult>, RecalibrationTables>.Cont
             .named("finalizeRecalTables")
             .of(new DoFn<RecalibrationTables, RecalibrationTables>() {
```
Comment (davidadamsphd) [inline src/main/java/org/broadinstitute/hellbender/dev/pipelines/bqsr/BaseRecalibratorDataflowUtils.java]: If this is going to be a common pattern, we could create a DoFnWLog that puts the logging stuff in the startBundle and finishBundle.

Comment (davidadamsphd) [inline src/main/java/org/broadinstitute/hellbender/dev/pipelines/bqsr/BaseRecalibratorDataflowUtils.java]: If this is going to be a common pattern, we could create a DoFnWLog that puts the logging stuff in the startBundle and finishBundle.

**Comment by davidadamsphd** [comment]
This is great! I'm glad you're making the tools we need to dig into performance.

A few things:
(1) Have you tested this? If so, how?
(2) Is there any issues with BunnyLog not being Serializable? It is used within processElement, so I'd expect it'd need to be.
(3) Please add some unit tests. I was thinking something like a fake program that just had delays (sleep) and verifying that the times produced were at least those you'd expect with the delays.
(4) I expect that this will eventually be paired with a logs analysis script. Do you have specific plans yet? I'd love to see a doc explaining what logging you'd like to see and how you think we should analyze it.

**Comment by jean-philippe-martin** [comment]
Thanks for having a look, David!
(1) I have tested it by running an instrumented BQSR and generating a report for it. I haven't checked in the reporting code yet - it's still very much in flux, and I'm not quite sure where to put it. Perhaps it should even be a separate project, since it's not hellbender-specific.
(2) This class is designed to *not* being transferred from one machine to another, so the timestamps are consistent. The idea is that a "START-END" span corresponds to a piece of work done on a single machine. It could also theoretically correspond to some "master" machine sending work and then getting answers back - but in both scenarios, all the log entries from the same operation come from the same machine.
(3) That would be testing log4j, not BunnyLog. I can try to find some way to add a test, sure.
(4) Yes, I started work on the analysis program already. It generates a report of how many times each operation was called, average call time, total CPU time. It also generates a visualization of when things are running. The latter turns out to be quite useful as we can see where there are gaps - indicating that some unlogged operation is taking longer than expected.

File: src/main/java/org/broadinstitute/hellbender/dev/pipelines/bqsr/BaseRecalibratorDataflowUtils.java (hunk 2 above)
Comment (jean-philippe-martin) [inline src/main/java/org/broadinstitute/hellbender/dev/pipelines/bqsr/BaseRecalibratorDataflowUtils.java]: OK, I've added one. To be maximally flexible I allow you to specify a different class and operation name. I also chose not to override processElement yet, though that's a matter of choice - perhaps later we may wish to override it to count the number of calls, perhaps time them automatically. Of course that may not make sense if processElement is very fast.

File: src/main/java/org/broadinstitute/hellbender/dev/pipelines/bqsr/BaseRecalibratorDataflowUtils.java [hunk 3]
```diff
@@ -114,28 +117,28 @@ public static void ensureReferenceIsReadable(final PipelineOptions popts, String
...
             .named("addQuantizationInfo")
-            .of(new DoFn<RecalibrationTables, BaseRecalOutput>() {
-              private static final long serialVersionUID = 1L;
-              @Override
-              public void processElement(ProcessContext c) {
-                RecalibrationTables rt = c.element();
-                BaseRecalibratorWorker baseRecalibratorWorker = BaseRecalibratorWorker.fromArgs(readsHeader, toolArgs);
-                baseRecalibratorWorker.onTraversalStart(null);
-                //BaseRecalOutput ret = new BaseRecalOutput(rt, baseRecalibratorWorker.getQuantizationInfo(rt), baseRecalibratorWorker.getRequestedCovariates());
+                    .of(new DoFnWLog<RecalibrationTables, BaseRecalOutput>("addQuantizationInfo") {
```
Comment (davidadamsphd) [inline src/main/java/org/broadinstitute/hellbender/dev/pipelines/bqsr/BaseRecalibratorDataflowUtils.java]: Weird whitespace bug?

**Comment by davidadamsphd** [comment]
(1) Sounds good.
(2) Ah, makes sense. I thought I saw the creation of the bunny outside of the scope of the worker (which it isn't)
(3) Yes and no. Suppose you switched/refactored from log4j to something else, you'd want to know that the timing still makes sense (or if log4j messed up their timing in a new release). It's not really an issue as log4j isn't going to mess up big and whatever you'd change to would also likely be as already tested as log4j. 
I'll leave it up to you if you want to add timing tests or not.
(4) Great! 
I think this work is something that we should keep the Dataflow team abreast of. There is no reason why great profiling tools couldn't eventually make their way into the SDK code and Google Console.

LTGM. Feel free to merge once my whitespace comment is addressed (with or without any more tests).

File: src/main/java/org/broadinstitute/hellbender/dev/pipelines/bqsr/BaseRecalibratorDataflowUtils.java (hunk 3 above)
Comment (jean-philippe-martin) [inline src/main/java/org/broadinstitute/hellbender/dev/pipelines/bqsr/BaseRecalibratorDataflowUtils.java]: argh, Idea is causing trouble again.

**Comment by jean-philippe-martin** [comment]
OK, that should be good. I'm going to squish, rebase, all that good stuff.

**Comment by jean-philippe-martin** [comment]
There you go, a clean rebased commit. That maintains the previous whitespace.

**Comment by davidadamsphd** [comment]
LGTM again

**Pull request merged**